tools/src/main/java/org/apache/kafka/tools/ManifestWorkspace.java

```diff
@@ -0,0 +1,573 @@
+                new URI("jar", writableJar.toUri().toString(), ""),
+                Collections.emptyMap()
+        )) {
+            Path zipRoot = jar.getRootDirectories().iterator().next();
+            rewriteClassHierarchyManifest(dryRun, zipRoot, manifestState);
+        } catch (URISyntaxException e) {
+            throw new IOException(e);
+        }
+        if (Files.exists(writableJar) && jarIsEmpty(writableJar)) {
+            Files.delete(writableJar);
+        }
+    }
+
+    private void createJar(Path path) throws IOException {
+        Objects.requireNonNull(path, "path must be non-null");
+        try (ZipOutputStream stream = new ZipOutputStream(Files.newOutputStream(
+                path,
+                StandardOpenOption.CREATE,
+                StandardOpenOption.WRITE,
+                StandardOpenOption.TRUNCATE_EXISTING
+        ))) {
+            stream.closeEntry();
+        }
+    }
+
+    private boolean jarIsEmpty(Path path) throws IOException {
+        Objects.requireNonNull(path, "path must be non-null");
+        try (ZipInputStream stream = new ZipInputStream(Files.newInputStream(
+                path,
+                StandardOpenOption.READ
+        ))) {
+            return stream.getNextEntry() == null;
+        }
+    }
+
+    /**
+     * Rewrite a multiple manifest files contained inside a class hierarchy.
+     * Will create the files and parent directories if they not exist.
+     * Will delete the files and parent directories within the plugin location if they become empty.
+     *
+     * @param dryRun         True if the rewrite should be applied, false if it should be simulated.
+     * @param pluginLocation Path to top-level of class hierarchy for a plugin
+     * @param manifestState  Map from plugin type to Class names of plugins which should appear in that manifest
+     */
+    private void rewriteClassHierarchyManifest(boolean dryRun, Path pluginLocation, Map<PluginType, Set<String>> manifestState) throws IOException, TerseException {
+        Objects.requireNonNull(pluginLocation, "pluginLocation must be non-null");
+        Objects.requireNonNull(manifestState, "manifestState must be non-null");
+        if (!Files.exists(pluginLocation)) {
+            throw new TerseException(pluginLocation + " does not exist");
+        }
+        if (!Files.isWritable(pluginLocation)) {
+            throw new TerseException(pluginLocation + " is not writable");
+        }
+        for (Map.Entry<PluginType, Set<String>> manifest : manifestState.entrySet()) {
+            PluginType type = manifest.getKey();
+            Set<String> elements = manifest.getValue();
+            Path manifestPath = pluginLocation.resolve(MANIFEST_PREFIX).resolve(type.superClass().getName());
+            rewriteManifestFile(dryRun, manifestPath, elements);
+        }
+    }
+
+    /**
+     * Rewrite a file, creating it and any parent directories if it does not exist, deleting it if specified
```

Comment:
What does "deleting it if specified" mean? I would think that'd refer to a method parameter but there is no, e.g., `boolean delete` here.

Reply:
This was nonsense. This (and other places where deletes happen) should only delete something if the script would otherwise leave it empty. An empty manifest will be removed, an empty directory will be removed, and an empty jar file will be removed.

I changed it so that `rewriteManifestFile` is only responsible for creating and deleting manifest files, `rewriteClassHierarchyManifest` is responsible for creating and deleting META-INF and services directories. `rewriteJar` was already responsible for creating and deleting jar files.